README.md
HV4D
====



Implementation of [HV4D] (https://eden.dei.uc.pt/~cmfonsec/guerreiro-cccg2012.pdf), an algorithm to compute the hypervolume indicator in four dimensions, in quadratic time. Minimization is assumed.


**Note**: Although only *nondominated* points that strongly dominate the reference point contribute to the hypervolume, the code is prepared to deal with all other points, including *repeated* points, which are all ignored (in the case of repeated points, one of the copies is considered, and the remaining ones are skipped). Warnings will be raised if any of the points do not strongly dominate the reference point. Moreover, different points may have (some) equal coordinates.



License
--------
	

Except where indicated otherwise in individual source files, this software is Copyright © 2011, 2017 Andreia P. Guerreiro.

This program is free software. You can redistribute it and/or modify it under the terms of the GNU General Public License, version 3, as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

Appropriate reference to this software should be made when describing research in which it played a substantive role, so that it may be replicated and verified by others. The algorithm which this software implements is described in detail in [1]. 



Building
--------


In GNU/Linux, the program can be compiled from source by invoking:

    make

It is recommended that you compile it specifically for your architecture. Depending on the compiler and version of the compiler you use, there are different ways to achieve this. For recent GCC versions, make will pick a suitable -march argument based on the processor of the build machine. This can be overridden by passing a `MARCH=` argument to make. Similarly if you use the Intel C compiler, it will pick a sensible default architecture (`-xHOST`) for you. If you want to override this, pass `XARCH=` to `make`. So, to build for an Intel *Core2* with the GCC compiler, you would use:

    make MARCH=core2

For the Intel C compiler you should use:

    make XARCH=SSSE3

Generally, `make` will try to pick good flags for you, but you can override them by passing an `OPT_CFLAGS` argument to `make` if you wish. To build an unoptimized version of `hv4d` you could run:

    make OPT_CFLAGS="-O0 -g"

Finally, if you do not want to see the command line of each compiler invocation, pass `S=1` to make.





General Usage
-------------------

**SYNOPSIS** 

    hv4d [OPTIONS] [FILE...]
    
**DESCRIPTION**

Compute the hypervolume indicator of the data set(s) in FILE(s).

With no FILE, read from the standard input.

**COMMAND LINE OPTIONS**

	 -h, --help          print this summary and exit.                          
	     --version       print version number and exit.                        
	 -v, --verbose       print some information (time, coordinate-wise maximum 
		             	 and minimum, etc)                                     
	 -q, --quiet         print only the hypervolume (as opposed to --verbose). 
	 -u, --union         treat all input sets within a FILE as a single set.   
	 -r, --reference=POINT use POINT as the reference point. POINT must be within  
		                 quotes, e.g., "10 10 10". If no reference point is  
		                 given, it is taken as the coordinate-wise maximum of  
		                 all input points.                                     
	 -s, --suffix=STRING Create an output file for each input file by appending
		                 this suffix. This is ignored when reading from stdin. 
		                 If missing, output is sent to stdout.    



Detailed Usage
-------------------

The program reads sets of points in the file(s) specified in the command line:

    ./hv4d data

or standard input:

    cat data | ./hv4d

In input files, each point is given in a separate line, and point coordinates in each line are separated by whitespace. An empty line, or a line beginning with a  hash sign (#), denotes a separate set.


Sets in an input file may be treated as a single set by using option `-u`: 

    ./hv4d data -u


The reference point can be set by giving option `-r`.

    ./hv4d -r "10 10 10 10" data

 If no reference point is given, the default is the coordinate-wise maximum of all input points in all files.

For the other options available, check the output of `./hv4d --help`.

    



Examples
--------

**Input File(s)**

Empty lines and lines starting with a hash sign (#) at the beginning and/or at the end of the file are ignored.

Example of valid content of input files:

    1   1   4   4
    4   4   1   1
    
    2   2   3   3
    3   3   2   2

Another example:

    #
    6 4 9 6
    7 3 7 1
    8 2 3 2
    9 1 2 4
    #
    1 9 5 4
    2 8 3 1
    #
    3 7 8 1
    4 6 3 9
    5 5 5 8
    #
    


**Compilation**

Example of a basic compilation:

    make march=corei7


**Execution**

The code can be tested with any of the data sets provided in folder `examples`. After compilation, run:

    ./hv4d examples/exampleInput.inp -r "1.1 1.1 1.1 1.1" 
    

The hypervolume indicator of each of the 2 data sets in `exampleInput.inp ` given the reference point `(1.1, 1.1, 1.1, 1.1)` will be printed in separate lines, in the same order as the point sets in the input file:

	0.936963193064112
	0.861528076506543

By running:


    ./hv4d examples/exampleInput.inp -r "1.1 1.1 1.1 1.1" -u
    
the result will be the hypervolume indicator of the union of the two sets in the file:

	1.05104266084533
    
    


Other sources
-------------

Version 1.1 of this code is also integrated in [PyGMO](http://esa.github.io/pygmo/index.html), and can be found [here](https://github.com/esa/pagmo/tree/master/src/util/hv_algorithm/hv4d_cpp_original).


References
----------

[1] Andreia P. Guerreiro, Carlos M. Fonseca, and Michael T. M. Emmerich. "A fast dimension-sweep algorithm for the hypervolume indicator in four dimensions". In CCCG, pages 77–82, 2012. [[pdf](https://eden.dei.uc.pt/~cmfonsec/guerreiro-cccg2012.pdf)]

[2] Andreia P. Guerreiro. Efficient algorithms for the assessment of stochastic multiobjective optimizers. Master’s thesis, IST, Technical University of Lisbon, Portugal, 2011. [[pdf](https://eden.dei.uc.pt/~cmfonsec/AndreiaGuerreiroMSc.pdf)]



CHANGELOG.md
# Change Log

<!-- ## [Unreleased] -->

## [1.2] - 2017-02-07
### Fixed
- Points that do not dominate the reference point could lead to incorrect results.

### Changed
- Raise a warning instead of an error if no point strongly dominates the reference point.

### Added
- The CHANGELOG.md file.

## [1.1] - 2017-01-15
### Added
- Support for dominated solutions (these are ignored).

## 1.0 - 2011
- First implementation of HV4D.


[1.1]: https://github.com/apguerreiro/HV4D/releases/tag/v1.1
[1.2]: https://github.com/apguerreiro/HV4D/releases/tag/v1.2


hv4d.c
/*************************************************************************

 hv4d - hypervolume indicator computation in 4D

 ---------------------------------------------------------------------

                       Copyright (c) 2011, 2017
                Andreia P. Guerreiro <apg@dei.uc.pt>

 This program is free software (software libre); you can redistribute
 it and/or modify it under the terms of the GNU General Public License,
 version 3, as published by the Free Software Foundation.

 This program is distributed in the hope that it will be useful, but
 WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, you can obtain a copy of the GNU
 General Public License at:
                 http://www.gnu.org/copyleft/gpl.html
 or by writing to:
           Free Software Foundation, Inc., 59 Temple Place,
                 Suite 330, Boston, MA 02111-1307 USA


*************************************************************************/

/*************************************************************************

 References:

 ---------------------------------------------------------------------
 [1] Andreia P. Guerreiro, Carlos M. Fonseca, and Michael T. M. Emmerich.
    A fast dimension-sweep algorithm for the hypervolume indicator in four
    dimensions. In CCCG, pages 77–82, 2012.

 [2] Andreia P. Guerreiro. Efficient algorithms for the assessment
    of stochastic multiobjective optimizers. Master’s thesis, IST,
    Technical University of Lisbon, Portugal, 2011.


*************************************************************************/



#include "timer.h"
#include <stdlib.h>
#include <stdio.h>
#include <limits.h>
#include <float.h>


#if __GNUC__ >= 3
# define __hv_unused    __attribute__ ((unused))
#else
# define __hv_unused    /* no 'unused' attribute available */
#endif



typedef struct dlnode {
  double x[4];                    /* The data vector              */
  struct dlnode *nexty;
  struct dlnode *prevy;
  struct dlnode *nextz;
  struct dlnode *prevz;
  double xrightbelow; //x value of the closest point (according to x) with lower y and z values
} dlnode_t;


//box
typedef struct bx{
    double lx, ly, lz;
    double ux, uy, uz;

    struct bx * next;
    struct bx * prev;
}box;

//box list
typedef struct bxl{
    box * head;
    box * tail;
}boxlist;



static int compare_node(const void *p1, const void* p2)
{
    const double x1 = *((*(const dlnode_t **)p1)->x+3);
    const double x2 = *((*(const dlnode_t **)p2)->x+3);

    return (x1 < x2) ? -1 : (x1 > x2) ? 1 : 0;
}


static int strongly_dominates(double * x1, const double * ref){
    int i;
    for(i = 0; i < 4 && x1[i] < ref[i]; i++);
    return (i == 4);
}


/*
 * Setup circular double-linked list in each dimension
 * and exclude points that do not strongly dominate the reference point.
 * size will return the number of points that strongly dominate the reference
 */
static dlnode_t *
setup_cdllist(double *data, int * size, const double *ref)
{
    int d = 4;
    dlnode_t *head;
    dlnode_t **scratch;
    int i, j;

    int n = *size;
    head  = malloc ((n+1) * sizeof(dlnode_t));


    int lasti = 0;
    for (i = 0; i < n; i++) {
        
        if(strongly_dominates(data + i*d, ref)){
            for(j = 0; j < d; j++){
                head[lasti].x[j] = data[i*d + j];
            }
            lasti++;
        }
    }
    n = lasti; //shrink n in case some points do not strongly dominate ref
    
    head[n].x[0] = head[n].x[1] = head[n].x[2] = -DBL_MAX;
    head[n].x[3] = ref[3];


    scratch = malloc(n * sizeof(dlnode_t*));
        
    for (i = 0; i < n; i++){
        scratch[i] = head + i;
    }

    qsort(scratch, n, sizeof(dlnode_t*), compare_node);

    double * data2 = (double *) malloc (n * 4 * sizeof(double));
    for(i = 0; i < n; i++){
        for(j = 0; j < d; j++){
            data2[i*d + j] = scratch[i]->x[j];
        }
    }


    for(i = 0; i < n; i++){
        for(j = 0; j < d; j++)
            head[i].x[j] = data2[i*d + j];
    }

    *size = n;
    
    free(data2);

    free(scratch);

    return head;
}

static void free_cdllist(dlnode_t * head)
{
    free(head);
}

static void free_sentinels(dlnode_t * list){

        free(list->nexty);
        free(list->prevy);

        free(list->nextz);
        free(list->prevz);
        
}


//note: it is not necessary to define uz
static box * new_box(double lx, double ly, double lz, double ux, double uy, double uz){
    box * b = (box *)malloc(sizeof(box));
    
    b->lx = lx;
    b->ly = ly;
    b->lz = lz;
    
    b->ux = ux;
    b->uy = uy;
    b->uz = uz;
    
    return b;
}

//push_left == push_head
static void push_left(boxlist * bl, box * b){
    b->next = bl->head;
    bl->head->prev = b;
    b->prev = NULL;

    bl->head = b;
}


//may be empty
static void push_left_e(boxlist * bl, box * b){
    if(bl->head == NULL){
        bl->head = b;
        bl->tail = b;

        b->next = NULL;
        b->prev = NULL;

    }else{

        b->next = bl->head;
        bl->head->prev = b;
        b->prev = NULL;

        bl->head = b;
    }
}


//box b is freed and its volume is returned
static double close_box(box *b){
        double volume = (b->ux - b->lx) * (b->uy - b->ly) * (b->uz - b->lz);
        free(b);

        return volume;
}

static double shrink_box(box *b, double ux){
        double volume = (b->ux - ux) * (b->uy - b->ly) * (b->uz - b->lz);
        b->ux = ux;
        return volume;
}

static double close_boxes_right(boxlist *bl, double x, double uz){
    double volume = 0;	
    box * b = bl->tail;
    while(b->lx > x){
        bl->tail = b->prev;
        bl->tail->next = NULL;
        b->uz = uz;
        volume += close_box(b);
        b = bl->tail;
    }
    //b = bl->tail;
        if(x < b->ux){
        b->uz = uz;
        volume += shrink_box(b, x);
        
    }

    return volume;

}

static double close_boxes_left(boxlist *bl, double y, double uz){

    if(y >= bl->head->uy)
        return 0;

    double volume = 0;	
    box * b = bl->head;

    double lastlx = b->ux;
    double lx = b->lx;
    double ly = b->ly;
    double lz = b->lz;

    while(b != NULL && b->uy >= y){

        lastlx = b->ux;
        b->uz = uz;
        b->ly = y;
        bl->head = b->next;
        
        volume += close_box(b);
        b = bl->head;
    }

    if(bl->head != NULL)
        bl->head->prev = NULL;
    else
        bl->tail = NULL;

    b = new_box(lx, ly, lz, lastlx, y, 0);
    push_left_e(bl, b);

    return volume;

}


static double close_all_boxes(boxlist *bl, double uz){
    double volume = 0;
    box * b = bl->head;
    while(b != NULL){
        b->uz = uz;
        bl->head = b->next;
        volume += close_box(b);
        b = bl->head;
    }
    
    bl->head = bl->tail = NULL;
    return volume;
}



static void insert_after_y(dlnode_t *new, dlnode_t *prev){
    
    new->prevy = prev;
    new->nexty = prev->nexty;
    new->nexty->prevy = new;
    prev->nexty = new;

}

static void insert_after_z(dlnode_t *new, dlnode_t *prev){
    
    new->prevz = prev;
    new->nextz = prev->nextz;
    new->nextz->prevz = new;
    prev->nextz = new;

}


static void remove_y(dlnode_t *p){
    
    p->prevy->nexty = p->nexty;
    p->nexty->prevy = p->prevy;
    
}

static void remove_z(dlnode_t *p){
    
    p->prevz->nextz = p->nextz;
    p->nextz->prevz = p->prevz;
    
}


static boxlist * init_box_list(dlnode_t *p, dlnode_t * ynext){

    
    double previous_x = p->xrightbelow;

    dlnode_t * q = ynext; // y is swept in ascending order

    boxlist * bl = (boxlist *)malloc(sizeof(boxlist));
    box * b = (box *)malloc(sizeof(box));
    
    b->prev = b->next = NULL;
    bl->head = bl->tail = b; //box b (sentinel) will be removed later

    //box limits in x and y are defined (vertical boxes)
    while(q->x[0] > p->x[0] || q->x[2] > p->x[2]){

        // =
        if(q->x[2] <= p->x[2] && q->x[0] < previous_x && q->x[0] > p->x[0]){
            b = new_box(q->x[0], p->x[1], p->x[2], previous_x, q->x[1], 0);
            push_left(bl, b);

            previous_x = q->x[0];    
        }
        
        q = q->nexty;
    }
    
    b = new_box(p->x[0], p->x[1], p->x[2], previous_x, q->x[1], 0);
    push_left(bl, b);
    
    
    //the box created in the beginning is removed
    bl->tail = bl->tail->prev;
    free(bl->tail->next);
    bl->tail->next = NULL;

    
    return bl;
    
}


static double update_boxes(boxlist * bl, dlnode_t *p, dlnode_t * znext){
    
    double volume = 0;
    dlnode_t * q;
    
    q = znext;

    while(bl->head != NULL){      
            if(q->x[0] <= p->x[0]){

                if(q->x[1] <= p->x[1]){ // p is dominated in x and y

                    //close all boxes
                    //update volume                 
                    //last iteration
                     volume += close_all_boxes(bl, q->x[2]);
                }else{
                    
                    //update z
                    //close boxes in the left
                    //update volume
                     volume += close_boxes_left(bl, q->x[1], q->x[2]);
                }

                if(q->x[1] >= p->x[1] && p->x[0] < q->xrightbelow){
                    q->xrightbelow = p->x[0];
                }

                
            }else{ //since q does not dominate p and p does not dominate q, q has to
                    //to be to the right of p and below p
                

                //update z
                //closes boxes in the right
                //update volume
                 volume += close_boxes_right(bl, q->x[0], q->x[2]);

            }

        q = q->nextz;

    }

    free(bl);
    
    return volume;
    
}


//corresponds to HV4D - contribution function of the pseudocode
static double
hv4d_increment3DA(dlnode_t *p, dlnode_t * ynext, dlnode_t *znext){

     boxlist * bl = init_box_list(p, ynext);
//      if(bl == NULL)
//          return 0;

    return update_boxes(bl, p, znext);
}



//inner cycle (sweeps points according to coordinate z)
//note: returns the contribution of p only
//Function responsible for keeping points sorted according to coordinates y and z,
//for updating xrightbelow and for removing dominated points
static double
hv4d_increment3D(dlnode_t *list, dlnode_t *p, const double * ref){
    double loss = 0, gain;
    dlnode_t * q = list->prevz;
    dlnode_t * yprev = list->nexty;
    double xrightbelow = ref[0];
    dlnode_t * zprev =  list->nextz;
    dlnode_t * head3 = list->nextz;

    while(q != head3){

        if(q->x[2] >= p->x[2] && q->x[1] >= p->x[1] && q->x[0] >= p->x[0]){
            remove_y(q);
            remove_z(q);
            loss += hv4d_increment3DA(q, q->nexty, q->nextz);
            
        }else{

            if(q->x[2] > zprev->x[2] && (q->x[2] < p->x[2] || (q->x[2] == p->x[2] &&  q->x[1] <= p->x[1]))){
                 //because of xrightbelow, in order to avoid height 0 boxes (in spite of not causing wrong results)
                zprev = q;
            }

            if(q->x[1] > yprev->x[1] && (q->x[1] < p->x[1] || (q->x[1] == p->x[1] &&  q->x[2] <= p->x[2]))){ //testequal.4d.6
                yprev = q;
            }
            
            if(q->x[2] <= p->x[2] && q->x[1] <= p->x[1]){
 
              if(q->x[0] <= p->x[0]){
                return 0;
              }else if(q->x[0] < xrightbelow){ //q->x[0] > p->x[0]
                xrightbelow = q->x[0];
              }
            }
        }
        q = q->prevz;
    }


    p->xrightbelow = xrightbelow;


    gain = hv4d_increment3DA(p, yprev->nexty, zprev->nextz);

    //add p to the lists of coordinates two and three
    insert_after_z(p, zprev);
    insert_after_y(p, yprev);

    return gain - loss;
}


//outer cycle (sweeps list L_4)
static double
hv4dA(dlnode_t *list, int c, const double * ref)
{
    dlnode_t *p = list;
    double hyperv = 0; //hvol4D
    double hypera = 0; //hvol3D
        
    int i;
    for (i = 0; i < c; i++) {
        hypera = hypera + hv4d_increment3D(list+c, p, ref);
        hyperv = hyperv + hypera * ((p+1)->x[3] - p->x[3]);
        p++;
    }

    return hyperv;
}






static void add_sentinels(dlnode_t * list, const double * ref){
    
    int i;
    
    for(i = 0; i <= 1; i++){
        dlnode_t * sentineli, * sentinelf;
        sentineli = (dlnode_t *) malloc(sizeof(dlnode_t));
        sentinelf = (dlnode_t *) malloc(sizeof(dlnode_t));

        sentineli->xrightbelow = -DBL_MAX;
        sentinelf->xrightbelow = -DBL_MAX;

        
        if(i == 0){
            sentineli->nexty = sentinelf;
            sentinelf->prevy = sentineli;
            list->nexty = sentineli;
            sentineli->prevy = list;
            list->prevy = sentinelf;
            sentinelf->nexty = list;
            
            
            sentineli->x[0] = ref[0];
            sentineli->x[1] = -DBL_MAX;
            sentineli->x[2] = -DBL_MAX;
            
            sentinelf->x[0] = -DBL_MAX;
            sentinelf->x[1] = ref[1];
            sentinelf->x[2] = -DBL_MAX;
        }else{
            sentineli->nextz = sentinelf;
            sentinelf->prevz = sentineli;
            list->nextz = sentineli;
            sentineli->prevz = list;
            list->prevz = sentinelf;
            sentinelf->nextz = list;
            
            sentineli->x[0] = DBL_MAX;
            sentineli->x[1] = DBL_MAX;
            sentineli->x[2] = -DBL_MAX;
            
            sentinelf->x[0] = -DBL_MAX;
            sentinelf->x[1] = -DBL_MAX;
            sentinelf->x[2] = ref[2];
            
        }

        sentineli->x[3] = DBL_MAX;
        sentinelf->x[3] = DBL_MAX;

        
    }
    
}


double hv4d(double *data, int n, const double *ref)
{
    dlnode_t *list;
    double hyperv;
    int nv = n;

    if (n == 0) { 
        hyperv = 0.0;
    } else {

        list = setup_cdllist(data, &nv, ref);
        //nv <= n (number of points that strongly dominate ref)
        
        add_sentinels(list+nv, ref);
        
        hyperv = hv4dA(list, nv, ref);
        
        free_sentinels(list+nv);
        /* Clean up.  */
        free_cdllist (list);
      
    }
    
    return hyperv;
}



main-hv4d.c
/*************************************************************************

 hv4d: main program

 ---------------------------------------------------------------------

                        Copyright (c) 2011, 2017
                Andreia P. Guerreiro <apg@dei.uc.pt>
             
                       Copyright (c) 2010
                  Carlos M. Fonseca <cmfonsec@dei.uc.pt>
             Manuel Lopez-Ibanez <manuel.lopez-ibanez@ulb.ac.be>
                    Luis Paquete <paquete@dei.uc.pt>
                Andreia P. Guerreiro <apg@dei.uc.pt>

 This program is free software (software libre); you can redistribute
 it and/or modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2 of the
 License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful, but
 WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, you can obtain a copy of the GNU
 General Public License at:
                 http://www.gnu.org/copyleft/gpl.html
 or by writing to:
           Free Software Foundation, Inc., 59 Temple Place,
                 Suite 330, Boston, MA 02111-1307 USA


*************************************************************************/
#include "io.h"
#include "hv4d.h"
#include "timer.h"

#include <errno.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <math.h>
#include <ctype.h> // for isspace()

#include <unistd.h>  // for getopt()
#include <getopt.h> // for getopt_long()

#ifdef __USE_GNU
extern char *program_invocation_short_name;
#else
char *program_invocation_short_name;
#endif

static const char * const stdin_name = "<stdin>";
static int verbose_flag = 1;
static bool union_flag = false;
static char *suffix = NULL;

static void usage(void)
{
    printf("\n"
           "Usage: %s [OPTIONS] [FILE...]\n\n", program_invocation_short_name);

    printf(
"Calculate the hypervolume of each input set of each FILE. \n"
"With no FILE, or when FILE is -, read standard input.\n\n"

"Options:\n"
" -h, --help          print this summary and exit.                          \n"
"     --version       print version number and exit.                        \n"
" -v, --verbose       print some information (time, coordinate-wise maximum \n"
"                     and minimum, etc)                                     \n"
" -q, --quiet         print only the hypervolume (as opposed to --verbose). \n"
" -u, --union         treat all input sets within a FILE as a single set.   \n"
" -r, --reference=POINT use POINT as the reference point. POINT must be within\n"
"                     quotes, e.g., \"10 10 10\". If no reference point is  \n"
"                     given, it is taken as the coordinate-wise maximum of  \n"
"                     all input points.                                     \n"
/*
"                     given, it is taken as max + 0.1 * (max - min) for each\n"
"                     coordinate from the union of all input points.        \n"
*/
" -s, --suffix=STRING Create an output file for each input file by appending\n"
"                     this suffix. This is ignored when reading from stdin. \n"
"                     If missing, output is sent to stdout.                 \n"
"\n");

}

static void version(void)
{
    printf("%s version " VERSION
#ifdef ARCH
           " (optimised for " ARCH ")"
#endif
           "\n\n", program_invocation_short_name);

    printf(
"Copyright (c) 2011, 2017"
"\nAndreia P. Guerreiro <apg@dei.uc.pt>\n"
"\nCopyright (c) 2010"
"\nCarlos M. Fonseca <cmfonsec@dei.uc.pt>"
"\nManuel Lopez-Ibanez <manuel.lopez-ibanez@ulb.ac.be>"
"\nLuis Paquete <paquete@dei.uc.pt>"
"\nAndreia P. Guerreiro <apg@dei.uc.pt>\n"
"\n"
"This is free software, and you are welcome to redistribute it under certain\n"
"conditions.  See the GNU General Public License for details. There is NO   \n"
"warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"
"\n"        );
}

static inline void
vector_printf (const double *vector, int size)
{
    int k;
    for (k = 0; k < size; k++)
        printf (" %f", vector[k]);
}

static double *
read_reference(char * str, int *nobj)
{
    double * reference;
    char * endp;
    char * cursor;

    int k = 0, size = 10;

    reference = malloc(size * sizeof(double));
    endp = str;

    do {
        cursor = endp;
        if (k == size) {
            size += 10;
            reference = realloc(reference, size * sizeof(double));
        }
        reference[k] = strtod(cursor, &endp);
        k++;
    } while (cursor != endp);

    // not end of string: error
    while (*cursor != '\0') {
        if (!isspace(*cursor)) return NULL;
        cursor++;
    }

    // no number: error
    if (k == 1) return NULL;

    *nobj = k-1;
    return reference;
}

static inline void
handle_read_data_error (int err, const char *filename)
{
    switch (err) {
    case 0: /* No error */
        break;

    case READ_INPUT_FILE_EMPTY:
        errprintf ("%s: no input data.", filename);
        exit (EXIT_FAILURE);

    case READ_INPUT_WRONG_INITIAL_DIM:
        errprintf ("check the argument of -r, --reference.\n");
    default:
        exit (EXIT_FAILURE);
    }
}

static void
data_range (double **maximum, double **minimum,
            const double *data, int nobj, int rows)
{
    int n, k = 0;
    int r = 0;

    if (*maximum == NULL) {
        *maximum = malloc (nobj*sizeof(double));
        for (k = 0; k < nobj; k++)
            (*maximum)[k] = data[k];
        r = 1;
    }

    if (*minimum == NULL) {
        *minimum = malloc (nobj*sizeof(double));
        for (k = 0; k < nobj; k++)
            (*minimum)[k] = data[k];
        r = 1;
    }

    for (; r < rows; r++) {
        for (n = 0; n < nobj; n++, k++) {
            if ((*maximum)[n] < data[k])
                (*maximum)[n] = data[k];
            if ((*minimum)[n] > data[k])
                (*minimum)[n] = data[k];
        }
    }
}

static void
file_range (const char *filename, double **maximum_p, double **minimum_p,
            int *dim_p)
{
    double *data = NULL;
    int *cumsizes = NULL;
    int nruns = 0;
    int dim = *dim_p;
    double *maximum = *maximum_p;
    double *minimum = *minimum_p;

    handle_read_data_error (
        read_data (filename, &data, &dim, &cumsizes, &nruns), filename);

    data_range (&maximum, &minimum, data, dim, cumsizes[nruns-1]);

    *dim_p = dim;
    *maximum_p = maximum;
    *minimum_p = minimum;

    free (data);
    free (cumsizes);
}

/*
   FILENAME: input filename. If NULL, read stdin.

   REFERENCE: reference point. If NULL, use MAXIMUM.

   MAXIMUM: maximum objective vector. If NULL, caculate it from the
   input file.

   NOBJ_P: pointer to number of objectives. If NULL, calculate it from
   input file.

*/

static void
hv_file (const char *filename, double *reference,
         double *maximum, double *minimum, int *nobj_p)
{
    double *data = NULL;
    int *cumsizes = NULL;
    int cumsize;
    int nruns = 0;
    int n;
    int nobj = *nobj_p;
    char *outfilename = NULL;
    FILE *outfile = stdout;
    bool setmax = false;
    bool setref = false;

    int err = read_data (filename, &data, &nobj, &cumsizes, &nruns);
    if (!filename) filename = stdin_name;
    handle_read_data_error (err, filename);

    if (filename != stdin_name && suffix) {
        int outfilename_len = strlen(filename) + strlen(suffix) + 1;

        outfilename = malloc (sizeof(char) * outfilename_len);
        strcpy (outfilename, filename);
        strcat (outfilename, suffix);

        outfile = fopen (outfilename, "w");
        if (outfile == NULL) {
            errprintf ("%s: %s\n", outfilename, strerror(errno));
            exit (EXIT_FAILURE);
        }
    }

    if (union_flag) {
        cumsizes[0] = cumsizes[nruns - 1];
        nruns = 1;
    }

    if (verbose_flag >= 2)
        printf("# file: %s\n", filename);

    if (maximum == NULL) {
        setmax = true;
        data_range (&maximum, &minimum, data, nobj, cumsizes[nruns-1]);
        if (verbose_flag >= 2) {
            printf ("# maximum:");
            vector_printf (maximum, nobj);
            printf ("\n");
            printf ("# minimum:");
            vector_printf (minimum, nobj);
            printf ("\n");
        }
    }

    if (reference != NULL) {
        for (n = 0; n < nobj; n++) {
            if (reference[n] <= maximum[n]) {
                warnprintf ("%s: some points do not strictly dominate "
                            "the reference point and they will be discarded",
                            filename);
                break;
            }
        }
    } else {
        setref = true;
        reference = malloc(nobj * sizeof(double));
//         double v;
//         char str[20];
        for (n = 0; n < nobj; n++) {
            /* default reference point is: *//*
            sprintf(str, "%f", maximum[n]);
//             v = (double)(index(str, '.')-&str[0]);
            v= strlen(str);*/
//             printf("v: %d\n", v);
//             v -= 16;
            reference[n] = maximum[n] + maximum[n] * pow(10.0,-15);
//             printf("maximum = %16.15g, m = %16.15g\n", maximum[n], reference[n]);
            /* however, a better reference point is:
            reference[n] = maximum[n] + 0.1 * (maximum[n] - minimum[n]);
            so that extreme points have some influence. */
        }
    }

    if (verbose_flag >= 2) {
        printf ("# reference:");
        vector_printf (reference, nobj);
        printf ("\n");
    }

    for (n = 0, cumsize = 0; n < nruns; cumsize = cumsizes[n], n++) {
        double time_elapsed_cpu;
        double volume;

        if (verbose_flag >= 2)
            fprintf (outfile, "# Data set %d:\n", n + 1);


        Timer_start ();
        volume = hv4d (&data[nobj * cumsize],
                            cumsizes[n] - cumsize, reference);
        if (volume <= 0.0) {
            warnprintf ("%s: none of the points strongly dominates the reference point",
                            filename);
//             errprintf ("none of the points strictly dominates the reference point\n");
//             exit (EXIT_FAILURE);
        }

        time_elapsed_cpu = Timer_elapsed_virtual ();

        fprintf (outfile, "%-16.15g\n", volume);



        if (verbose_flag >= 2) {
            fprintf (outfile, "# Time computing hypervolume (cpu): %f seconds\n", time_elapsed_cpu);
        }

    }

    if (outfilename) {
        if (verbose_flag)
            fprintf (stderr, "# %s -> %s\n", filename, outfilename);
        fclose (outfile);
        free (outfilename);
    }
    free (data);
    free (cumsizes);
    if (setmax){
        free (maximum);
        free (minimum);
    }
    if (setref) free(reference);
    *nobj_p = nobj;
}

int main(int argc, char *argv[])
{
    double *reference = NULL;
    int nobj = 0;
    int numfiles, k;

    int opt; /* it's actually going to hold a char.  */
    int longopt_index;

    /* see the man page for getopt_long for an explanation of these fields.  */
    static struct option long_options[] = {
        {"help",       no_argument,       NULL, 'h'},
        {"version",    no_argument,       NULL, 'V'},
        {"verbose",    no_argument,       NULL, 'v'},
        {"quiet",      no_argument,       NULL, 'q'},
        {"reference",  required_argument, NULL, 'r'},
        {"union",      no_argument,       NULL, 'u'},
        {"suffix",     required_argument, NULL, 's'},
#ifdef EXPERIMENTAL
        {"reorder",    no_argument,       NULL, 'R'},
#endif

        {NULL, 0, NULL, 0} /* marks end of list */
    };

#ifndef __USE_GNU
    program_invocation_short_name = argv[0];
#endif

    while (0 < (opt = getopt_long (argc, argv, "hVvquRr:s:",
                                   long_options, &longopt_index))) {
        switch (opt) {
        case 'r': // --reference
            reference = read_reference (optarg, &nobj);
            if (reference == NULL) {
                errprintf ("invalid reference point '%s",
                           optarg);
                exit (EXIT_FAILURE);
            }
            break;

        case 'u': // --union
            union_flag = true;
            break;

        case 's': // --suffix
            suffix = optarg;
            break;

        case 'V': // --version
            version();
            exit(EXIT_SUCCESS);

        case 'q': // --quiet
            verbose_flag = 0;
            break;

        case 'v': // --verbose
            verbose_flag = 2;
            break;

        case '?':
            // getopt prints an error message right here
            fprintf (stderr, "Try `%s --help' for more information.\n",
                     program_invocation_short_name);
            exit(EXIT_FAILURE);
        case 'h':
            usage();
            exit(EXIT_SUCCESS);

        default: // should never happen
            abort();
        }
    }

    numfiles = argc - optind;

    if (numfiles < 1) /* Read stdin.  */
        hv_file (NULL, reference, NULL, NULL, &nobj);

    else if (numfiles == 1) {
        hv_file (argv[optind], reference, NULL, NULL, &nobj);
    }
    else {
        double *maximum = NULL;
        double *minimum = NULL;
        if (reference == NULL) {
            /* Calculate the maximum among all input files to use as
               reference point.  */
            for (k = 0; k < numfiles; k++)
                file_range (argv[optind + k], &maximum, &minimum, &nobj);

            if (verbose_flag >= 2) {
                printf ("# maximum:");
                vector_printf (maximum, nobj);
                printf ("\n");
                printf ("# minimum:");
                vector_printf (minimum, nobj);
                printf ("\n");
            }
        }
        for (k = 0; k < numfiles; k++)
            hv_file (argv[optind + k], reference, maximum, minimum, &nobj);
        
        free (maximum);
        free (minimum);
    }
    if (reference != NULL) free (reference);
    return EXIT_SUCCESS;
}
